File: src/Layout/widget/controller/io/Outputs.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package Layout.widget.controller.io;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.function.Consumer;

import org.reactfx.Subscription;

import com.google.common.reflect.TypeToken;

import static util.reactive.Util.maintain;

/**
 *
 * @author Plutonium_
 */
public class Outputs {
    private final Map<String,Output<?>> m;


    public Outputs() {
        m = new HashMap<>();
    }

    public <T> Output<T> create(UUID id, String name, TypeToken<? super T> type, T val) {
        Output<T> o = create(id, name, type.getRawType(), val);
        o.typet = type;
        return o;
    }

    public <T> Output<T> create(UUID id, String name, Class<? super T> type, T val) {
        Output<T> o = new Output<>(id,name,type);
                  o.setValue(val);
        m.put(name, o);
        return o;
    }

    public int getSize() {
        return m.size();
    }

    public Output getOutput(String name) {
        return m.get(name);
    }

    public boolean contains(String name) {
        return m.containsKey(name);
    }

    public boolean contains(Output i) {
         return m.containsValue(i); // slow impl./ w can rely on the below:
//        return m.containsKey(i.id.name); // doesnt guarantee correctness
    }

    public Collection<Output<?>> getOutputs() {
        return m.values();
    }

    public <T> Subscription monitor(String name, Consumer<T> action) {
        Output<T> o = (Output<T>)m.get(name);
        if(o==null) return Subscription.EMPTY;

        return maintain(o.val, action);
    }



}

File: src/sp/it/pl/layout/widget/controller/io/Outputs.kt
package sp.it.pl.layout.widget.controller.io

import org.reactfx.Subscription
import sp.it.pl.util.reactive.sync
import sp.it.pl.util.type.isSuperclassOf
import java.util.HashMap
import java.util.UUID

class Outputs {
    private val m = HashMap<String, Output<*>>()

    inline fun <reified T> create(id: UUID, name: String, value: T?) = create<T>(id, name, T::class.java, value)

    fun <T> create(id: UUID, name: String, type: Class<in T>, value: T?): Output<T?> {
        val o = Output(id, name, type)
        o.value = value
        m[name] = o
        return o
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> findOutput(type: Class<T>, name: String): Output<T?>? {
        val i = m[name]
        if (i!=null && type.isSuperclassOf(i.type)) throw ClassCastException()
        return i as Output<T?>?
    }

    fun <T> getOutput(type: Class<T>, name: String): Output<T?> = findOutput(type, name)!!

    inline fun <reified T> findOutput(name: String) = findOutput(T::class.java, name)

    inline fun <reified T> getOutput(name: String) = findOutput(T::class.java, name)!!

    operator fun contains(name: String) = m.containsKey(name)

    operator fun contains(i: Output<*>) = m.containsValue(i) // m.containsKey(i.id.name) <--faster, but not correct

    fun getSize(): Int = m.size

    fun getOutputs(): Collection<Output<*>> = m.values

    @Suppress("IfThenToElvis", "UNCHECKED_CAST")
    fun <T> monitor(name: String, action: (T) -> Unit): Subscription {
        val o = m[name] as Output<T>?
        return if (o==null) Subscription.EMPTY else o.`val` sync action
    }

}